File: components/ambient-sdk/python-sdk/ambient_platform/session_watch.py
"""Session watch functionality for real-time streaming of session changes via gRPC."""

import asyncio
import ssl
from typing import AsyncIterator, Optional, Union
from urllib.parse import urlparse

import grpc
from grpc import aio

from .session import Session


class SessionWatchEvent:
    """Represents a real-time session change event."""
    
    def __init__(self, event_type: str, session: Optional[Session] = None, resource_id: str = ""):
        self.type = event_type
        self.session = session
        self.resource_id = resource_id
    
    def is_created(self) -> bool:
        """Returns True if this is a creation event."""
        return self.type == "CREATED"
    
    def is_updated(self) -> bool:
        """Returns True if this is an update event."""
        return self.type == "UPDATED"
    
    def is_deleted(self) -> bool:
        """Returns True if this is a deletion event."""
        return self.type == "DELETED"
    
    def __repr__(self) -> str:
        return f"SessionWatchEvent(type={self.type}, resource_id={self.resource_id})"


class SessionWatcher:
    """Provides real-time session event streaming via gRPC."""
    
    def __init__(self, client, timeout: Optional[float] = None):
        self._client = client
        self._timeout = timeout or 1800.0  # 30 minutes
        self._channel: Optional[aio.Channel] = None
        self._stub = None
        self._stream = None
        self._closed = False
    
    async def __aenter__(self):
        await self.start()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
    
    async def start(self):
        """Start the gRPC watch stream."""
        if self._closed:
            raise RuntimeError("Watcher is closed")
        
        # Create gRPC connection
        grpc_address = self._derive_grpc_address()
        
        # Determine if we need TLS
        if self._client._base_url.startswith("https://"):
            credentials = grpc.ssl_channel_credentials()
        else:
            credentials = grpc.insecure_channel_credentials()
        
        # Create async channel
        self._channel = aio.secure_channel(grpc_address, credentials) if credentials != grpc.insecure_channel_credentials() else aio.insecure_channel(grpc_address)
        
        # Import generated protobuf stubs (would need to generate Python stubs)
        # For now, we'll use a placeholder implementation
        from . import _grpc_stubs  # This would be generated
        
        self._stub = _grpc_stubs.SessionServiceStub(self._channel)
        
        # Create metadata for authentication
        metadata = [
            ("authorization", f"Bearer {self._client._token}"),
            ("x-ambient-project", self._client._project),
        ]
        
        # Start watch stream
        request = _grpc_stubs.WatchSessionsRequest()
        self._stream = self._stub.WatchSessions(
            request,
            metadata=metadata,
            timeout=self._timeout
        )
    
    async def watch(self) -> AsyncIterator[SessionWatchEvent]:
        """Stream session watch events."""
        if not self._stream:
            await self.start()
        
        try:
            async for event in self._stream:
                yield self._convert_event(event)
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.CANCELLED:
                return
            raise RuntimeError(f"Watch stream error: {e}")
    
    async def close(self):
        """Close the watcher and clean up resources."""
        self._closed = True
        if self._stream:
            self._stream.cancel()
        if self._channel:
            await self._channel.close()
    
    def _derive_grpc_address(self) -> str:
        """Convert HTTP base URL to gRPC address."""
        parsed = urlparse(self._client._base_url)
        host = parsed.hostname
        
        # For ambient-api-server, gRPC typically runs on port 4434
        port = 4434
        if parsed.port:
            # If the URL has a custom port, keep it
            port = parsed.port
        
        return f"{host}:{port}"
    
    def _convert_event(self, event) -> SessionWatchEvent:
        """Convert protobuf event to SDK event."""
        event_type = ""
        if event.type == 1:  # EVENT_TYPE_CREATED
            event_type = "CREATED"
        elif event.type == 2:  # EVENT_TYPE_UPDATED
            event_type = "UPDATED"
        elif event.type == 3:  # EVENT_TYPE_DELETED
            event_type = "DELETED"
        else:
            event_type = "UNKNOWN"
        
        # Convert session if present
        session = None
        if hasattr(event, 'session') and event.session:
            session = self._convert_session(event.session)
        
        return SessionWatchEvent(
            event_type=event_type,
            session=session,
            resource_id=event.resource_id if hasattr(event, 'resource_id') else ""
        )
    
    def _convert_session(self, proto_session) -> Session:
        """Convert protobuf Session to SDK Session."""
        # This would convert the protobuf session to our Session dataclass
        # For now, we'll create a basic conversion
        data = {
            "name": proto_session.name if hasattr(proto_session, 'name') else "",
            "id": "",
            "kind": "Session",
            "href": "",
        }
        
        # Set metadata from protobuf
        if hasattr(proto_session, 'metadata') and proto_session.metadata:
            meta = proto_session.metadata
            if hasattr(meta, 'id'):
                data["id"] = meta.id
            if hasattr(meta, 'kind'):
                data["kind"] = meta.kind
            if hasattr(meta, 'href'):
                data["href"] = meta.href
            if hasattr(meta, 'created_at'):
                data["created_at"] = meta.created_at.ToDatetime().isoformat()
            if hasattr(meta, 'updated_at'):
                data["updated_at"] = meta.updated_at.ToDatetime().isoformat()
        
        # Set optional session fields
        optional_fields = [
            "repo_url", "prompt", "created_by_user_id", "assigned_user_id",
            "workflow_id", "repos", "timeout", "llm_model", "llm_temperature",
            "llm_max_tokens", "phase", "project_id"
        ]
        
        for field in optional_fields:
            if hasattr(proto_session, field):
                value = getattr(proto_session, field)
                if value is not None:
                    data[field] = value
        
        # Handle timestamp fields
        if hasattr(proto_session, 'start_time') and proto_session.start_time:
            data["start_time"] = proto_session.start_time.ToDatetime().isoformat()
        if hasattr(proto_session, 'completion_time') and proto_session.completion_time:
            data["completion_time"] = proto_session.completion_time.ToDatetime().isoformat()
        
        return Session.from_dict(data)


# Add watch method to SessionAPI
def add_watch_to_session_api():
    """Monkey patch to add watch method to SessionAPI."""
    from ._session_api import SessionAPI
    
    def watch(self, timeout: Optional[float] = None) -> SessionWatcher:
        """Create a session watcher for real-time events."""
        return SessionWatcher(self._client, timeout=timeout)
    
    # Add the watch method
    SessionAPI.watch = watch

File: components/ambient-sdk/go-sdk/client/session_watch.go
// Watch functionality for Session API
// Implements real-time streaming of session changes via gRPC

package client

import (
	"context"
	"crypto/tls"
	"fmt"
	"io"
	"net"
	"net/url"
	"os"
	"strings"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/metadata"

	ambient_v1 "github.com/ambient-code/platform/components/ambient-api-server/pkg/api/grpc/ambient/v1"
	"github.com/ambient-code/platform/components/ambient-sdk/go-sdk/types"
)

const grpcDefaultPort = "9000"

var defaultOpenShiftPatterns = []string{"apps.rosa", "apps.ocp", "apps.openshift"}

// MessageWatcher streams session messages from a single session via gRPC.
type MessageWatcher struct {
	conn          *grpc.ClientConn
	msgs          chan *types.SessionMessage
	errors        chan error
	ctx           context.Context
	cancel        context.CancelFunc
	timeoutCancel context.CancelFunc
	done          chan struct{}
}

// Messages returns a channel of session messages as they arrive.
func (w *MessageWatcher) Messages() <-chan *types.SessionMessage {
	return w.msgs
}

// Errors returns a channel of stream errors.
func (w *MessageWatcher) Errors() <-chan error {
	return w.errors
}

// Done returns a channel closed when the watcher stops.
func (w *MessageWatcher) Done() <-chan struct{} {
	return w.done
}

// Stop closes the watcher and releases resources.
func (w *MessageWatcher) Stop() {
	if w.timeoutCancel != nil {
		w.timeoutCancel()
	}
	w.cancel()
	if w.conn != nil {
		_ = w.conn.Close()
	}
}

func (w *MessageWatcher) receive(stream grpc.ServerStreamingClient[ambient_v1.SessionMessage]) {
	defer close(w.done)
	defer close(w.msgs)
	defer close(w.errors)

	for {
		select {
		case <-w.ctx.Done():
			return
		default:
			pbMsg, err := stream.Recv()
			if err != nil {
				if err == io.EOF {
					return
				}
				select {
				case w.errors <- fmt.Errorf("watch stream error: %w", err):
				case <-w.ctx.Done():
				}
				return
			}
			msg := protoMsgToSessionMessage(pbMsg)
			select {
			case w.msgs <- msg:
			case <-w.ctx.Done():
				return
			}
		}
	}
}

// WatchSessionMessages opens a gRPC stream for real-time session message delivery.
// afterSeq sets the replay cursor — pass 0 to replay all messages from the beginning.
// Call Stop() on the returned watcher to cancel and clean up.
func (a *SessionAPI) WatchSessionMessages(ctx context.Context, sessionID string, afterSeq int64, opts *WatchOptions) (*MessageWatcher, error) {
	if opts == nil {
		opts = &WatchOptions{Timeout: 30 * time.Minute}
	}

	conn, err := a.createGRPCConnection()
	if err != nil {
		return nil, fmt.Errorf("failed to create gRPC connection: %w", err)
	}

	grpcClient := ambient_v1.NewSessionServiceClient(conn)

	md := metadata.New(map[string]string{
		"authorization":     "Bearer " + a.client.token,
		"x-ambient-project": a.client.project,
	})

	watchCtx, watchCancel := context.WithCancel(ctx)
	watcher := &MessageWatcher{
		conn:   conn,
		msgs:   make(chan *types.SessionMessage, 64),
		errors: make(chan error, 5),
		ctx:    watchCtx,
		cancel: watchCancel,
		done:   make(chan struct{}),
	}

	streamCtx := metadata.NewOutgoingContext(watchCtx, md)
	if opts.Timeout > 0 {
		var timeoutCancel context.CancelFunc
		streamCtx, timeoutCancel = context.WithTimeout(streamCtx, opts.Timeout)
		watcher.timeoutCancel = timeoutCancel
	}

	stream, err := grpcClient.WatchSessionMessages(streamCtx, &ambient_v1.WatchSessionMessagesRequest{
		SessionId: sessionID,
		AfterSeq:  afterSeq,
	})
	// stream type: grpc.ServerStreamingClient[ambient_v1.SessionMessage]
	if err != nil {
		watchCancel()
		_ = conn.Close()
		return nil, fmt.Errorf("failed to start WatchSessionMessages stream: %w", err)
	}

	go watcher.receive(stream)

	return watcher, nil
}

func protoMsgToSessionMessage(pb *ambient_v1.SessionMessage) *types.SessionMessage {
	msg := &types.SessionMessage{
		SessionID: pb.GetSessionId(),
		Seq:       int(pb.GetSeq()),
		EventType: pb.GetEventType(),
		Payload:   pb.GetPayload(),
	}
	msg.ID = pb.GetId()
	if pb.GetCreatedAt() != nil {
		t := pb.GetCreatedAt().AsTime()
		msg.CreatedAt = &t
	}
	return msg
}

// SessionWatcher provides real-time session events
type SessionWatcher struct {
	stream        ambient_v1.SessionService_WatchSessionsClient
	conn          *grpc.ClientConn
	events        chan *types.SessionWatchEvent
	errors        chan error
	ctx           context.Context
	cancel        context.CancelFunc
	timeoutCancel context.CancelFunc
	done          chan struct{}
}

// WatchOptions configures session watching
type WatchOptions struct {
	// ResourceVersion to start watching from (empty = latest)
	ResourceVersion string
	// Timeout for the watch connection
	Timeout time.Duration
}

// Watch creates a new session watcher with real-time events
func (a *SessionAPI) Watch(ctx context.Context, opts *WatchOptions) (*SessionWatcher, error) {
	if opts == nil {
		opts = &WatchOptions{Timeout: 30 * time.Minute}
	}

	// Create gRPC connection to API server
	conn, err := a.createGRPCConnection()
	if err != nil {
		return nil, fmt.Errorf("failed to create gRPC connection: %w", err)
	}

	// Create session service client
	grpcClient := ambient_v1.NewSessionServiceClient(conn)

	// Add authentication metadata
	md := metadata.New(map[string]string{
		"authorization":     "Bearer " + a.client.token,
		"x-ambient-project": a.client.project,
	})

	// Create watcher with its own cancellable context
	watchCtx, watchCancel := context.WithCancel(ctx)
	watcher := &SessionWatcher{
		conn:   conn,
		events: make(chan *types.SessionWatchEvent, 10),
		errors: make(chan error, 5),
		ctx:    watchCtx,
		cancel: watchCancel,
		done:   make(chan struct{}),
	}

	// Apply timeout to stream context but store cancel on watcher so Stop() controls lifetime
	streamCtx := metadata.NewOutgoingContext(watchCtx, md)
	if opts.Timeout > 0 {
		var timeoutCancel context.CancelFunc
		streamCtx, timeoutCancel = context.WithTimeout(streamCtx, opts.Timeout)
		watcher.timeoutCancel = timeoutCancel
	}

	// Start watch stream
	stream, err := grpcClient.WatchSessions(streamCtx, &ambient_v1.WatchSessionsRequest{})
	if err != nil {
		watchCancel()
		_ = conn.Close()
		return nil, fmt.Errorf("failed to start watch stream: %w", err)
	}
	watcher.stream = stream

	// Start goroutine to receive events
	go watcher.receiveEvents()

	return watcher, nil
}

// Events returns a channel of session watch events
func (w *SessionWatcher) Events() <-chan *types.SessionWatchEvent {
	return w.events
}

// Errors returns a channel of watch errors
func (w *SessionWatcher) Errors() <-chan error {
	return w.errors
}

// Done returns a channel that's closed when the watcher stops
func (w *SessionWatcher) Done() <-chan struct{} {
	return w.done
}

// Stop closes the watcher and cleans up resources
func (w *SessionWatcher) Stop() {
	if w.timeoutCancel != nil {
		w.timeoutCancel()
	}
	w.cancel()
	if w.conn != nil {
		_ = w.conn.Close()
	}
}

// receiveEvents runs in a goroutine to receive and convert events
func (w *SessionWatcher) receiveEvents() {
	defer close(w.done)
	defer close(w.events)
	defer close(w.errors)

	for {
		select {
		case <-w.ctx.Done():
			return
		default:
			event, err := w.stream.Recv()
			if err != nil {
				if err == io.EOF {
					return // Stream ended normally
				}
				select {
				case w.errors <- fmt.Errorf("watch stream error: %w", err):
				case <-w.ctx.Done():
				}
				return
			}

			// Convert protobuf event to SDK event
			sdkEvent := w.convertEvent(event)
			if sdkEvent != nil {
				select {
				case w.events <- sdkEvent:
				case <-w.ctx.Done():
					return
				}
			}
		}
	}
}

// convertEvent converts protobuf SessionWatchEvent to SDK types
func (w *SessionWatcher) convertEvent(event *ambient_v1.SessionWatchEvent) *types.SessionWatchEvent {
	if event == nil {
		return nil
	}

	eventType := ""
	switch event.GetType() {
	case ambient_v1.EventType_EVENT_TYPE_CREATED:
		eventType = "CREATED"
	case ambient_v1.EventType_EVENT_TYPE_UPDATED:
		eventType = "UPDATED"
	case ambient_v1.EventType_EVENT_TYPE_DELETED:
		eventType = "DELETED"
	default:
		eventType = "UNKNOWN"
	}

	result := &types.SessionWatchEvent{
		Type:       eventType,
		ResourceID: event.GetResourceId(),
	}

	// Convert session if present
	if event.GetSession() != nil {
		result.Session = w.convertSession(event.GetSession())
	}

	return result
}

// convertSession converts protobuf Session to SDK Session
func (w *SessionWatcher) convertSession(session *ambient_v1.Session) *types.Session {
	if session == nil {
		return nil
	}

	result := &types.Session{
		Name: session.GetName(),
	}

	// Set metadata
	if meta := session.GetMetadata(); meta != nil {
		result.ID = meta.GetId()
		result.Kind = meta.GetKind()
		result.Href = meta.GetHref()
		if meta.GetCreatedAt() != nil {
			createdAt := meta.GetCreatedAt().AsTime()
			result.CreatedAt = &createdAt
		}
		if meta.GetUpdatedAt() != nil {
			updatedAt := meta.GetUpdatedAt().AsTime()
			result.UpdatedAt = &updatedAt
		}
	}

	// Set optional fields
	if session.RepoUrl != nil {
		result.RepoURL = *session.RepoUrl
	}
	if session.Prompt != nil {
		result.Prompt = *session.Prompt
	}
	if session.CreatedByUserId != nil {
		result.CreatedByUserID = *session.CreatedByUserId
	}
	if session.AssignedUserId != nil {
		result.AssignedUserID = *session.AssignedUserId
	}
	if session.WorkflowId != nil {
		result.WorkflowID = *session.WorkflowId
	}
	if session.Repos != nil {
		result.Repos = *session.Repos
	}
	if session.Timeout != nil {
		result.Timeout = int(*session.Timeout)
	}
	if session.LlmModel != nil {
		result.LlmModel = *session.LlmModel
	}
	if session.LlmTemperature != nil {
		result.LlmTemperature = *session.LlmTemperature
	}
	if session.LlmMaxTokens != nil {
		result.LlmMaxTokens = int(*session.LlmMaxTokens)
	}
	if session.Phase != nil {
		result.Phase = *session.Phase
	}
	if session.ProjectId != nil {
		result.ProjectID = *session.ProjectId
	}
	if session.GetStartTime() != nil {
		startTime := session.GetStartTime().AsTime()
		result.StartTime = &startTime
	}
	if session.GetCompletionTime() != nil {
		completionTime := session.GetCompletionTime().AsTime()
		result.CompletionTime = &completionTime
	}

	return result
}

// createGRPCConnection creates a gRPC connection to the ambient-api-server
func (a *SessionAPI) createGRPCConnection() (*grpc.ClientConn, error) {
	grpcAddr := a.deriveGRPCAddress()

	var creds credentials.TransportCredentials
	if strings.HasPrefix(a.client.baseURL, "https://") {
		tlsCfg := &tls.Config{MinVersion: tls.VersionTLS12}
		if a.client.insecureSkipVerify {
			tlsCfg.InsecureSkipVerify = true //nolint:gosec
		}
		creds = credentials.NewTLS(tlsCfg)
	} else {
		creds = insecure.NewCredentials()
	}

	conn, err := grpc.NewClient(grpcAddr, grpc.WithTransportCredentials(creds))
	if err != nil {
		return nil, fmt.Errorf("failed to create gRPC client at %s: %w", grpcAddr, err)
	}

	return conn, nil
}

// deriveGRPCAddress converts HTTP base URL to gRPC address
func (a *SessionAPI) deriveGRPCAddress() string {
	// Allow explicit override via environment variable
	if grpcURL := os.Getenv("AMBIENT_GRPC_URL"); grpcURL != "" {
		return grpcURL
	}

	u, err := url.Parse(a.client.baseURL)
	if err != nil || u.Host == "" {
		return net.JoinHostPort(a.client.baseURL, grpcDefaultPort)
	}

	// Configurable OpenShift Route pattern detection
	if isOpenShiftRoute(u.Host) {
		return deriveOpenShiftGRPCAddress(u)
	}

	// Use the hostname only (strip any HTTP port) and apply gRPC default port
	return net.JoinHostPort(u.Hostname(), grpcDefaultPort)
}

// isOpenShiftRoute detects if the hostname follows OpenShift Route patterns
func isOpenShiftRoute(host string) bool {
	patterns := defaultOpenShiftPatterns
	if customPattern := os.Getenv("AMBIENT_OPENSHIFT_PATTERN"); customPattern != "" {
		patterns = []string{customPattern}
	}

	for _, pattern := range patterns {
		if strings.Contains(host, pattern) && strings.Contains(host, "ambient-api-server") {
			return true
		}
	}
	return false
}

// deriveOpenShiftGRPCAddress converts OpenShift HTTP route to gRPC route
func deriveOpenShiftGRPCAddress(u *url.URL) string {
	// Convert: ambient-api-server-namespace.apps.rosa.xxx
	// To:      ambient-api-server-grpc-namespace.apps.rosa.xxx
	grpcHost := strings.Replace(u.Host, "ambient-api-server", "ambient-api-server-grpc", 1)

	// Use port 443 for OpenShift Route (maps to pod port 9000 via targetPort)
	// OpenShift Routes only expose ports 80/443 externally
	return grpcHost + ":443"
}
